src/tpm2.c

```diff
@@ -290,103 +300,68 @@ static int TPM2_CalcHmac(TPM2_CTX* ctx, int inHandleCnt,
 }
 #endif /* !WOLFTPM2_NO_WOLFCRYPT */
 
-
-/* Send Command Wrapper */
-typedef enum CmdFlags {
-    CMD_FLAG_NONE = 0x00,
-    CMD_FLAG_DEC = 0x01, /* command request supports parameter encryption */
-    CMD_FLAG_ENC = 0x02, /* command response supports parameter encryption */
-} CmdFlags_t;
-
-#if 0
-typedef struct {
-    int authCnt;
-    int inHandleCnt;
-    int outHandleCnt;
-    CmdFlags_t flags;
-} CmdInfo_t;
-#endif
-
-static int TPM2_CommandProcess(TPM2_CTX* ctx, BYTE *cmd, UINT32 cmdSz,
-    int authCnt, int inHandleCnt, CmdFlags_t flags)
+static int TPM2_CommandProcess(TPM2_CTX* ctx, TPM2_Packet* packet,
+    CmdInfo_t* info)
 {
     int rc = TPM_RC_SUCCESS;
     TPM_CC cmdCode;
-    TPM2_Packet packet;
-    UINT32 packetSz, authSz;
-    BYTE *param, *encParam;
-    int paramSz, encParamSz, authPos, i;
+    UINT32 cmdSz, authSz;
+    BYTE *param, *encParam = NULL;
+    int paramSz, encParamSz = 0, authPos, i;
 
-    /* TODO: Check flags to avoid paramenc for command that don't support it */
-    (void)flags;
-
-    /* Parse packet size and TPM command code */
-    XMEMSET(&packet, 0, sizeof(packet));
-    packet.buf = cmd;
-    packet.size = cmdSz;
-
-    /* Skip tag */
-    packet.pos += sizeof(UINT16);
-    TPM2_Packet_ParseU32(&packet, &packetSz); /* Extract packet size */
-    TPM2_Packet_ParseU32(&packet, &cmdCode);  /* Extract TPM Command Code */
+    /* Parse header */
+    packet->pos = sizeof(UINT16); /* Skip tag */
+    TPM2_Packet_ParseU32(packet, &cmdSz);    /* Extract Command Size - total size including header */
+    TPM2_Packet_ParseU32(packet, &cmdCode);  /* Extract TPM Command Code */
 
     /* Skip the handles area */
-    packet.pos += (inHandleCnt * sizeof(TPM_HANDLE));
+    packet->pos += (info->inHandleCnt * sizeof(TPM_HANDLE));
 
     /* Parse Auth */
-    authPos = packet.pos; /* mark position for start of auth */
-    TPM2_Packet_ParseU32(&packet, &authSz);
-    packet.pos += authSz;
+    authPos = packet->pos; /* mark position for start of auth */
+    TPM2_Packet_ParseU32(packet, &authSz);
+    packet->pos += authSz;
     
     /* Mark parameter data */
-    param = &packet.buf[packet.pos];
-    paramSz = packet.size - packet.pos;
+    param = &packet->buf[packet->pos];
+    paramSz = cmdSz - packet->pos;
 
     /* Mark "first" encryption parameter */
-    encParam = param;
-    encParamSz = paramSz;
-    if (packet.pos + (int)sizeof(UINT16) <= packet.size) {
+    if (info->flags & CMD_FLAG_ENC2) {
         UINT16 tempSz;
-        /* Parse first parameter */
-        TPM2_Packet_ParseU16(&packet, &tempSz);
-        encParam += sizeof(UINT16);
+        TPM2_Packet_ParseU16(packet, &tempSz);
+        encParam = param + sizeof(UINT16);
+        encParamSz = tempSz;
+    }
+    else if (info->flags & CMD_FLAG_ENC4) {
+        UINT32 tempSz;
+        TPM2_Packet_ParseU32(packet, &tempSz);
```

Comment:
I think this should be also U16, but with offset of U16(+2).
Because I believe this is the case when multiple TPM2B parameters are wrapped as one. Still, the size field of the TPM2B structures is always UINT16. However, we sometimes must extract the inner size field, i.e the size field of the first wrapped parameter
tpm2b-struct{ 
tpm2b.size;
tpm2b.firstparam.size; < we want this
tpm2b.firstparam;
tpm2b.secondparam.size;
tpm2b.secondparam;
}